packages/angular/build/src/builders/karma/application_builder.ts

```diff
@@ -718,3 +707,83 @@ function getInstrumentationExcludedPaths(root: string, excludedPaths: string[]):
 
   return excluded;
 }
+function getBaseKarmaOptions(
+  options: NormalizedKarmaBuilderOptions,
+  context: BuilderContext,
+): ConfigOptions {
+  const singleRun = !options.watch;
+
+  // Determine project name from builder context target
+  const projectName = context.target?.project;
+  if (!projectName) {
+    throw new Error(`The 'karma' builder requires a target to be specified.`);
+  }
+
+  const karmaOptions: ConfigOptions = options.karmaConfig
+    ? {}
+    : getBuiltInKarmaConfig(context.workspaceRoot, projectName);
+
+  karmaOptions.singleRun = singleRun;
+
+  // Workaround https://github.com/angular/angular-cli/issues/28271, by clearing context by default
+  // for single run executions. Not clearing context for multi-run (watched) builds allows the
+  // Jasmine Spec Runner to be visible in the browser after test execution.
+  karmaOptions.client ??= {};
+  karmaOptions.client.clearContext ??= singleRun ?? false; // `singleRun` defaults to `false` per Karma docs.
```

Comment:
Does `?? false` ever apply? `singleRun` is a `const` that's assigned a boolean (`!expr`). I don't think the default value in Karma matters since we hard override it in 726?